core/src/queue.rs

```diff
@@ -137,59 +145,88 @@ impl Queue {
+                        tx,
+                        err: Error::Full,
+                    });
                 }
-                Entry::Vacant(entry) => entry,
+
+                // Insert entry first so that the `tx` popped from `queue` will always have a `(hash, tx)` record in `txs`.
+                entry.insert(tx);
+                let queue_to_push = if signature_check {
+                    &self.queue
+                } else {
+                    &self.waiting_buffer
+                };
+                queue_to_push.push(hash).map_err(|err_hash| {
+                    let (_, err_tx) = self
+                        .txs
+                        .remove(&err_hash)
+                        .expect("Inserted just before match");
+                    Failure {
+                        tx: err_tx,
+                        err: Error::Full,
+                    }
+                })
+            }
+        }
+    }
+
+    /// Pop single transaction from the signature buffer. Record unsigned transaction in `seen`.
+    fn pop_from_signature_buffer(
+        &self,
+        seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
+        wsv: &WorldStateView,
+    ) -> Option<VersionedAcceptedTransaction> {
+        loop {
+            let hash = self.waiting_buffer.pop()?;
+            let entry = match self.txs.entry(hash) {
+                Entry::Occupied(entry) => entry,
+                // FIXME: Reachable under high load. Investigate, see if it's a problem.
+                Entry::Vacant(_) => continue,
             };
 
-            // Reason for such insertion order is to avoid situation
-            // when poped from the `queue` hash does not yet has corresponding (hash, tx) record in `txs`
-            entry.insert(tx);
-            self.queue.push(hash).map_err(|err_hash| {
-                let (_, err_tx) = self
-                    .txs
-                    .remove(&err_hash)
-                    .expect("Inserted just before match");
-                (err_tx, Error::Full)
-            })
+            match self.check_tx(entry.get(), wsv) {
+                Ok(MustUse(signature_check)) => {
+                    // Transactions are not removed from the queue until expired or committed
+                    seen.push(hash);
+                    if signature_check {
+                        return Some(entry.get().clone());
+                    }
+                }
+                Err(_) => {
+                    entry.remove_entry();
+                }
+            }
         }
     }
 
-    /// Pop single transaction.
-    ///
-    /// Records unsigned transaction in `seen`.
-    #[allow(
-        clippy::expect_used,
-        clippy::unwrap_in_result,
-        clippy::cognitive_complexity
-    )]
-    fn pop(
+    /// Pop single transaction from the queue. Record unsigned transaction in `seen`.
+    fn pop_from_queue(
         &self,
         seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
```

Comment:
`seen` seems useless now? How can you have an unsigned transaction in the queue?

Reply:
Right, seen is indeed useless in case of `queue`.

Reply:
I will update documentation, because we record poped transactions in `seen` as well in order to put then back into the queue.